pyacq/core/rpc/client.py

```diff
@@ -1,89 +1,428 @@
+        
+        logger.info("RPC connect to %s", address.decode())
+        self._socket.connect(address)
+        self.next_request_id = 0
+        self.futures = weakref.WeakValueDictionary()
+        
+        with RPCClient.clients_by_thread_lock:
+            RPCClient.clients_by_thread[key] = self
+        
+        # proxies generated by this client will be assigned these default options
+        self.default_proxy_options = {}
+        
+        self.connect_established = False
+        self.establishing_connect = False
+        self._disconnected = False
+
+        # Proxies we have received from other machines. 
+        self.proxies = {}
+
+        # For unserializing results returned from servers. This cannot be
+        # used to send proxies of local objects unless there is also a server
+        # for this thread..
+        self.serializer = MsgpackSerializer(client=self)
+        
+        self.ensure_connection()
+
+    def disconnected(self):
+        """Boolean indicating whether the server has disconnected from the client.
+        """
+        if self._disconnected:
+            return True
+        
+        # check to see if we have received any new messages
+        self._read_and_process_all()
+        return self._disconnected
+
+    def send(self, action, opts=None, return_type='auto', sync='sync', timeout=10.0):
+        """Send a request to the remote process.
+        
+        Parameters
+        ----------
+        action : str
+            The action to invoke on the remote process.
+        opts : None or dict
+            Extra options to be sent with the request. Each action requires a
+            different set of options.
+        return_type : 'auto' | 'proxy'
+            If 'proxy', then the return value is sent by proxy. If 'auto', then
+            the server decides based on the return type whether to send a proxy.
+        sync : str
+            If 'sync', then block and return the result when it becomes available.
+            If 'async', then return a Future instance immediately.
+            If 'off', then ask the remote server NOT to send a response and
+            return None immediately.
+        timeout : float
+            The amount of time to wait for a response when in synchronous
+            operation (sync='sync').
+        """
+        if self.disconnected():
+            raise RuntimeError("Cannot send request; server has already disconnected.")
+        cmd = {'action': action, 'return_type': return_type, 
+               'opts': opts}
+        if sync == 'off':
+            req_id = None
+        else:
+            req_id = self.next_request_id
+            self.next_request_id += 1
+        cmd['req_id'] = req_id
+        logger.info("RPC request '%s' to %s [req_id=%s]", cmd['action'], 
+                    self.address.decode(), req_id)
+        logger.debug("    => %s", cmd)
+        
+        # double-serialize opts to ensure that cmd can be read even if opts
```

Comment:
Let's say I put an object into `opts` that causes an error during unserialization. This would cause `req_id` to become unreadable, which would prevent sending back an error message. So the double serialization guarantees that we at least get `req_id` before moving on to the more dangerous `opts`. However, we can probably accomplish the same using multipart messages (and it might be faster as well)...  I will investigate.